components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/impl/DefaultProvisioningHandler.java

```diff
@@ -718,7 +719,42 @@ private String getUserStoreDomain(String userStoreDomain, UserRealm realm)
      * @return
      */
     protected String generatePassword() {
-        return RandomStringUtils.randomNumeric(12);
+
+        // Pick from some letters that won't be easily mistaken for each other.
+        // So, for example, omit o O and 0, 1 l and L.
+        // This will generate a random password which satisfy the following regex.
+        // ^((?=.*\\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%&*])).{12}$}
+        Random secureRandom = new SecureRandom();
+        String characters = "23456789abcdefghjkmnpqrstuvwxyzABCDEFGHJKMNPQRSTUVWXYZ!@#$%&*";
+        String digits = "23456789";
+        String lowercaseLetters = "abcdefghjkmnpqrstuvwxyz";
+        String uppercaseLetters = "ABCDEFGHJKMNPQRSTUVWXYZ";
+        String specialCharacters = "!@#$%&*";
+        int passwordLength = 12;
+        int mandatoryCharactersCount = 4;
+
+        StringBuilder pw = new StringBuilder();
+        int index;
+        for (int i = 0; i < passwordLength - mandatoryCharactersCount; i++) {
+            index = secureRandom.nextInt(characters.length());
+            pw.append(characters.charAt(index));
+        }
+
+        index = secureRandom.nextInt(digits.length());
+        pw.append(digits.charAt(index));
+
+        index = secureRandom.nextInt(lowercaseLetters.length());
+        pw.append(lowercaseLetters.charAt(index));
+
+        index = secureRandom.nextInt(uppercaseLetters.length());
+        pw.append(uppercaseLetters.charAt(index));
+
+        index = secureRandom.nextInt(specialCharacters.length());
+        pw.append(specialCharacters.charAt(index));
+
+        char[] password = new char[pw.length()];
+        pw.getChars(0, pw.length(), password, 0);
```

Comment:
How about shuffling the characters in the password to eliminate any predictable patterns?

Reply:
Since we already using SecureRandom objects I don't think it is necessary to shuffle again. In other password generation places also we are using the same logic (https://github.com/wso2/carbon-identity-framework/blob/bd6aa21bcb2b529584b5a8699437c9a6166378b0/components/user-mgt/org.wso2.carbon.user.mgt.common/src/main/java/org/wso2/carbon/user/mgt/common/DefaultPasswordGenerator.java#L33-L68).